Implement check of performance metrics after each chunk

paparici opened the pull request:
This pull request introduces a new **performance monitoring** for **Autosubmit**, enabling the **computation** and **notification** (using **e-mail**) of **performance** **metrics** for **jobs** if they aren't within a threshold ([issue](https://github.com/BSC-ES/autosubmit/issues/2462)). The changes include the addition of a base performance class,  specific implementations for job types and projects (e.g. DestinE). It's important to indicate that only is implemented the metric `SYPD`. 

An example of the **new configuration** for the performance:  
```
PERFORMANCE:
- PROJECT: 'DESTINE'
- SECTION: 
  - SIM
  - POST
- NOTIFY_ON: 
 - COMPLETED 
 - FAILED 
- NOTIFY_TO: 
  - e-mail1 
  - e-mail2
```

The configuration lets the user to select the jobs to compute and notify performance metrics. Moreover, if a project has specific performance metrics, they will be computed.  In the implementation, I did a dummy `DestinESIMPerformance` to show how it could be implemented (e.g. changes the threshold for SYPD, new metrics...).

**Question**: I would like to ask how to add integration tests (the unit tests are completed) for this implementation and where should I include it (tests for platforms.py?).

**P.S**: I'm looking for the threshold of **SYPD** for **DESTINE**. So, I would like to ask to have reviewed the code at the meantime, therefore I will need an additional commit to update the threshold.

paparici committed:
Implement performance metrics calculation for SIM jobs

- Added PerformanceFactory to create performance metric calculators based on job type.
- Implemented SIMPerformance class to compute SYPD (Seconds per Year per Day) metrics.
- Created SIMDestinEPerformance class inheriting from SIMPerformance for specific job metrics.
- Integrated performance metrics computation into the Platform class during job log recovery.
- Developed unit tests for performance metrics calculations, including edge cases and threshold checks.
- Added utility functions for SYPD calculations and under-performance percentage calculations.
- Implemented email notification for performance alerts when metrics fall below defined thresholds.

paparici committed:
Enhance docstrings for performance-related classes and methods to improve clarity and consistency

paparici committed:
Add to the try-except of _compute_performance_metrics the creation of the performance manager

MissterP commented:
@VindeeR told me to check for pylint errors. So, I will check that.

paparici committed:
Applied changes to reduce pylint errors.

MissterP commented on autosubmit/performance/base_performance.py:
```diff
@@ -35,17 +36,20 @@ class PerformanceMetricInfo(NamedTuple):
-    _mail_notifier = MailNotifier(BasicConfig()) # Default MailNotifier with BasicConfig 
+    _mail_notifier = MailNotifier(
+        BasicConfig()
+    )  # Default MailNotifier with BasicConfig
```
Hi! I tried to use **black** to reduce pylint warnings. I just saw that in .pylintrc states that max_size is 120 so, I will do the changes. Thank you for the information! 

I would like to ask if is there a standard tool for solving pylint warnings in Autosubmit?

kinow replied:
Not yet. We have an issue to add that, but we haven't had time yet. For now, the best option is to try to lint new code, leaving code not related to your change unchanged (easier to review :+1: ) (unless you are adding docs, mypy, or fixing code that's hard to read)

paparici committed:
Refactor to have max line size to 120.

paparici committed:
- Added integration tests for the performance computation and the new function of test mail.
- Modify the fixture "_create_autosubmit_exp" to accept re-write experiment configuration (Bruno helped).
- Added the performance section in the AutosubmitCnfig of platforms.py.
- Corrected how to check the status of a job in the performance factory and do a deep copy of the Autosubmit configuration.

MissterP commented:
Hi! It's not necessary to run the tests. It's still a work in progress. I just want to update my branch with the changes of master.

paparici committed:
Update configuration imports from performance.

paparici committed:
Enhance performance configuration handling and add chunk parameters

- Introduced chunk_length and chunk_unit properties in Job class.
- Updated BasePerformance and PerformanceFactory to use performance_config instead of AutosubmitConfig.
- Improved error handling for performance configuration in PerformanceFactory.
- Refactored SIMPerformance to utilize new chunk parameters.
- Added update_job_information method in Platform for better job configuration management.
- Updated integration and unit tests to mock performance configuration.

paparici committed:
Reorder job information update to ensure it occurs after performance manager creation

kinow commented:
> Hi! Still working on some issues detected from the integration tests. I would like to ask why the start_timestamp of a job could be greater than the finish_timestamp (executing tests in local) even when in the configuration of the job, the scripts states sleep 5. Moreover, I'm running the tests with the parameter -n 0.

Hmmm, weird. Could be, a) `sleep` is mocked, b) bug in our logic, c) something is taking too long or running too fast in the tests. We will try to replicate it later, thanks for the detailed info, Pablo :+1:

paparici committed:
- Correct the unit tests with the new changes to solve th integration tests.
-  Correct the usage of BasicConfig in BasePerformance.
- Emhance mock of MailNotifier in integration tests.

paparici committed:
Refactor chunk_size handling to use integers instead of strings and update related tests

paparici committed:
Refactor error management in SYPD computation to ensure timestamps are integers and improve validation

paparici committed:
Refactor to add the MailNotifier Instance to the Performance managers using the Inyection pattern.

kinow commented on test/unit/performance/utils.py:
```diff
@@ -0,0 +1,66 @@
+    """
+
+    @staticmethod
+    def get_SYPD(start_timestamp: int, finish_timestamp: int, chunk_size: str, chunk_size_unit: str) -> float:
```
As this class contains only static functions, it probably make more sense to make these functions, without a class, and then import then when/if needed.

The naming also needs to be lowercase (uppercase in Python is rarely used, reserved for constants [which are also normally not constants.... but that's just Python :shrug: ]).

MissterP replied:
On it! Thanks for the info!

kinow commented on test/unit/performance/utils.py:
```diff
@@ -0,0 +1,66 @@
+
+        :param threshold: The threshold value to compare against.
+        :type threshold: float
+
```
In Python docstrings, parameters and types, etc, do not need spaces between them.

MissterP replied:
On it! Thanks for the info!

kinow commented on test/unit/performance/utils.py:
```diff
@@ -0,0 +1,66 @@
+    @staticmethod
+    def get_under_performance(value: float, threshold: float) -> float:
+        """
+        Calculate the percentage under performance based on the value and threshold.
```
You can place this comment right after the `"""`, https://peps.python.org/pep-0008/#documentation-strings

That's like the header of the docs. It's normally short, very brief, just stating what it does in a few chars. Anything else in the lines below, with one blank line separating paragraphs.

kinow commented on autosubmit/performance/type_job/SIM/SIM_performance.py:
Python files normally do not have upper case, folders too are kept in lower case (unless we have a good/strong reason for doing so).

MissterP replied:
On it! Thanks for the info!

MissterP replied:
> Python files normally do not have upper case, folders too are kept in lower case (unless we have a good/strong reason for doing so).

Before starting to apply lint, PEP8, etc. I wanted to check that the integration tests are successful. However, I really appreciate your comments.

kinow reviewed (commented):
> Hi @kinow!
>
>I did a refactor to use the Injection pattern so that I can get the MailNotifier instance from a utils.py from my performance folder. However, I have the same issue were the patch isn't applying because it's being applied after the imports.

Do you mean in `mock_base_performance_notifier`? Could you point which part you need help with? I'm a bit lost as there are quite a few files in this PR :grimacing:

MissterP commented:
> Python files normally do not have upper case, folders too are kept in lower case (unless we have a good/strong reason for doing so).

Before starting to apply lint, PEP8, etc. I wanted to check that the integration tests are successful. However, I really appreciate your comments!
